engine/crates/partial-caching/src/output/shapes/building.rs

```diff
@@ -0,0 +1,355 @@
+    }
+
+    let nocache_partition_root = {
+        let (subset, selection_set) = plan.nocache_partition();
+        let selections = selection_set
+            .map(|selection| MergedSelection {
+                selection,
+                propagate_defer_label: None,
+            })
+            .collect();
+
+        build_output_shape(&mut objects, selections, subset)
+    };
+
+    OutputShapes {
+        objects,
+        cache_partition_roots,
+        nocache_partition_root,
+    }
+}
+
+fn build_output_shape(
+    objects: &mut Vec<ObjectShapeRecord>,
+    selections: Vec<MergedSelection<'_>>,
+    subset: &QuerySubset,
+) -> super::ObjectShapeId {
+    let type_conditions = FragmentIter::new(&selections, subset)
+        .filter_map(|fragment| fragment.type_condition())
+        .collect::<IndexSet<_>>();
+
+    if type_conditions.is_empty() {
+        let field_shapes = field_shapes_for_typename(&selections, subset, objects, None);
+
+        insert_record(objects, ObjectShapeRecord::Concrete { fields: field_shapes })
+    } else {
+        let mut types = Vec::with_capacity(type_conditions.len() + 1);
+
+        types.push((None, field_shapes_for_typename(&selections, subset, objects, None)));
+
+        for typename in type_conditions {
+            types.push((
+                Some(typename.to_string()),
+                field_shapes_for_typename(&selections, subset, objects, Some(typename)),
+            ));
+        }
+
+        insert_record(objects, ObjectShapeRecord::Polymorphic { types })
+    }
+}
+
+fn field_shapes_for_typename(
+    selections: &[MergedSelection<'_>],
+    subset: &QuerySubset,
+    objects: &mut Vec<ObjectShapeRecord>,
+    typename: Option<&str>,
+) -> Vec<FieldRecord> {
+    let mut grouped_fields = IndexMap::new();
+
+    collect_fields(&mut grouped_fields, &mut vec![], selections, subset, typename);
+
+    let merged_fields = merge_selection_sets(grouped_fields, subset);
+
+    let mut field_shapes = vec![];
+
+    for field in merged_fields {
+        let mut child_object = None;
+        if !field.selections.is_empty() {
+            child_object = Some(build_output_shape(objects, field.selections, subset));
+        }
+        field_shapes.push(FieldRecord {
+            response_key: field.response_key.to_string(),
+            defer_label: field.defer_label.map(ToString::to_string),
+            child_object,
+        });
+    }
+
+    field_shapes
+}
+
+struct CollectedField<'a> {
+    field: FieldSelection<'a>,
+
+    /// The label of the defer this selection is in if any
+    defer_label: Option<&'a str>,
+}
+
+/// A defer aware implementation of CollectFields from the GraphQL spec:
+///
+/// Note that this doesn't process include or skip currently.  I think this
+/// is fine and we can leave that up to the actual GraphQL server, but may
+/// need revisited if that's wrong.
+///
+/// http://spec.graphql.org/October2021/#CollectFields()
+fn collect_fields<'a>(
```

Comment:
yeah, i'm not the biggest fan of the extra noise doing recursive functions "properly" like that adds.  if this was a public function I agree it'd be worth it, but since it's private I think I probably prefer doing this?

Reply:
up to you, I think on the opposite it reduces noise as:
- it makes it explicit it's recursive, and makes it more obvious that grouped_fields and the other mutable is state for this recursion
- isolates the mutable part into this function, rather than leaking those in the caller.

but that's a preference and I've understood it, so 🤷

Reply:
Well, my thinking is that this:

```
fn collect_fields<'a>(
     selections: &[MergedSelection<'a>],
     subset: &'a QuerySubset,
     typename: Option<&'a str>,
) -> IndexMap<&'a str, Vec<CollectedField<'a>>> { ... }

fn recursive_collect_fields<'a>(
    grouped_fields: &mut IndexMap<&'a str, Vec<CollectedField<'a>>>,
     defer_stack: &mut Vec<&'a str>,
     selections: &[MergedSelection<'a>],
     subset: &'a QuerySubset,
     typename: Option<&'a str>,
) { ... }
```

is objectively more code than this:

```
fn collect_fields<'a>(
    grouped_fields: &mut IndexMap<&'a str, Vec<CollectedField<'a>>>,
     defer_stack: &mut Vec<&'a str>,
     selections: &[MergedSelection<'a>],
     subset: &'a QuerySubset,
     typename: Option<&'a str>,
) { ... }
```

But as you say: this is just my preference in this instance.